ingest-asset-reconciler/src/main/scala/uk/gov/nationalarchives/Lambda.scala

```diff
@@ -39,34 +42,107 @@ class Lambda extends LambdaRunner[Input, StateOutput, Config, Dependencies] {
+      lazy val fileNameWithoutExtension = assetChild.name
+      val potentialAssetChildTitleOrFileName = assetChild.title.getOrElse("")
+      val assetChildTitleOrFileName = if potentialAssetChildTitleOrFileName.isEmpty then fileNameWithoutExtension else potentialAssetChildTitleOrFileName
+
+      val numOfDotsInTitleOrFileName = assetChildTitleOrFileName.count(_ == '.')
+      val numOfDotsInTitleOfCo = titleOfCo.count(_ == '.')
+      val differenceInNumberOfDots = numOfDotsInTitleOrFileName - numOfDotsInTitleOfCo
+
+      val (titleOfCoWithoutExtension, assetChildTitleOrFileNameWithoutExtension) =
+        if (numOfDotsInTitleOrFileName == numOfDotsInTitleOfCo) (titleOfCo, assetChildTitleOrFileName)
+        else if (differenceInNumberOfDots == 1) (titleOfCo, stripFileExtension(assetChildTitleOrFileName))
+        else if (abs(differenceInNumberOfDots) > 1) (titleOfCo, assetChildTitleOrFileName) // then let it fail the equality comparison below
+        else (stripFileExtension(titleOfCo), assetChildTitleOrFileName)
+
+      titleOfCoWithoutExtension == assetChildTitleOrFileNameWithoutExtension
+    }
+
+  private def verifyFilesInDdbAreInPreservica(
+      childrenForRepresentationType: List[FileDynamoTable],
+      bitstreamInfoPerContentObject: Seq[BitStreamInfo],
+      assetId: UUID,
+      representationType: RepresentationType
+  ) = {
+    val childrenThatDidNotMatchOnChecksum =
+      childrenForRepresentationType.filter { assetChild =>
+        val bitstreamWithSameChecksum = bitstreamInfoPerContentObject.find { bitstreamInfoForCo =>
+          assetChild.checksumSha256 == bitstreamInfoForCo.fixity.value &&
+          coTitleMatchesAssetChildTitle(bitstreamInfoForCo.potentialCoTitle, assetChild)
+        }
+
+        bitstreamWithSameChecksum.isEmpty
+      }
+
+    lazy val idsOfChildrenThatDidNotMatchOnChecksum = childrenThatDidNotMatchOnChecksum.map(_.id)
+
+    if (childrenThatDidNotMatchOnChecksum.isEmpty) StateOutput(wasReconciled = true, "")
+    else
+      StateOutput(
+        wasReconciled = false,
+        s"Out of the ${childrenForRepresentationType.length} files expected to be ingested for assetId '$assetId' with representationType $representationType, " +
+          s"a checksum and title could not be matched with a file on Preservica for: ${idsOfChildrenThatDidNotMatchOnChecksum.mkString(", ")}"
+      )
+  }
+
+  private def generateSnsMessage(dependencies: Dependencies, assetId: UUID, lockTableName: String, batchId: String) =
+    for {
+      items <- dependencies.dynamoDbClient.getItems[IngestLockTable, PartitionKey](
+        List(PartitionKey(assetId)),
+        lockTableName
+      )
+
+      attributes <- IO.fromOption(items.headOption)(
+        new Exception(s"No items found for assetId '$assetId' from batchId '$batchId'")
+      )
+
+      message <- IO.fromEither(decode[AssetMessage](attributes.message.replace('\'', '\"')))
```

Comment:
What's the call to `replace('\''...` for?

Reply:
The tests kept throwing this error `software.amazon.awssdk.thirdparty.jackson.core.JsonParseException: Unexpected character ('p' (code 112)): was expecting comma to separate Object entries`; it's got to do with the lock-table's mock response JSON. It has an issue with me using double quotes in the `message` (the `message` itself is a JSON, so json within a json) so I replaced the double quotes in the JSON `message` with single quotes and then did a `.replace` reversal right before the decoding as this was the only place I could add it where it wouldn't get the error.